### PR TITLE
Addition of Caltech101 and Caltech256 object detection Datasets

### LINKED ISSUE #128: [Dataset] Add Caltech Datasets for object detection
Please add the following Caltech Dataset as new _Image object detection_ datasets: 

- [ ] Caltech 101 : https://data.caltech.edu/records/mzrjq-6wc02
- [ ] Caltech 256 : https://data.caltech.edu/records/nyy15-4j048

Please ensure consistency with previous supported dataset with
- [ ] Reuse of the existing download and cache
- [ ] Equivalent test suite
- [ ] Documentation update

### PR DESCRIPTION
**Pull request by DerrickUnleashed**
### Fixes:

- Consistent with `cifar10_dataset()`
- Includes comprehensive test suite including dataloader tests.
- Adds detailed roxygen documentation for both datasets.
- There exists no `test/train` split for both datasets to maintain consistency with Pytorch documentation.
- Caltech101 uses `target_type` params which can take `category` , `annotation` or `all`.
  - for `category` the returned output of `first_item$y` is label name.
  - for `annotation` the returned output of `first_item$y` is a named list with `obj_contour` and `box_coord`.
  - for `all` the returned output of `first_item$y` is a named list with `label` , `obj_contour` and `box_coord`.

Closes #128

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -10,6 +10,8 @@
 - Added `emnist_dataset()` – a dataset loader for the EMNIST dataset. (#152, @DerrickUnleashed)
 - Added `fgvc_aircraft_dataset()` – a dataset loader for the FGCV-Aircraft dataset. (#156, @DerrickUnleashed)
 - add tiff image support to `folder_dataset()` (#169, @cregouby)
+- Added `caltech101_dataset()` – a dataset loader for the Caltech101 dataset. (#158, @DerrickUnleashed)
+- Added `caltech256_dataset()` – a dataset loader for the Caltech256 dataset. (#158, @DerrickUnleashed)
```
Comment (cregouby) [inline NEWS.md:14]: **improvement** a single line for both is enough.

File: R/dataset-caltech.R [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+#'
+#' @param root Character. Root directory for dataset storage. Default is a temporary directory. Data will be stored under `root/caltech101`.
+#' @param target_type Character. One of `"category"`, `"annotation"`, or `"all"`. Determines the type of target returned. Default is `"category"`.
+#' @param transform Optional function to apply to each image (e.g., resizing or normalization).
+#' @param target_transform Optional function to transform the target. Default is `NULL`.
+#' @param download Logical. Whether to download and extract the dataset if not already available. Default is `FALSE`.
+#'
+#' @return An object of class \code{caltech101_dataset}, which behaves like a torch dataset.
+#' Each element is a named list:
+#' - `x`: a H x W x 3 integer array representing an RGB image.
```
Comment (cregouby) [inline R/dataset-caltech.R:19]: **todo** all datasets returns x with channel first. Please apply a `aperm()` to for consistency.
**improvement** please use consistent image **shape description** with other datasets : 
https://github.com/mlverse/torchvision/blob/61c84b0d2b96385855996e4f35b4b3fb8ca33cd6/R/dataset-coco.R#L16
**praise** I like the "x" and "y" naming (see #165) 
**improvement** please use the **same names** and description (if relevant) as coco_dataset for y content
https://github.com/mlverse/torchvision/blob/61c84b0d2b96385855996e4f35b4b3fb8ca33cd6/R/dataset-coco.R#L19-L24

File: R/dataset-caltech.R [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+#' Loads the Caltech-101 dataset consisting of 101 object categories, with images of varied dimensions.
+#' Each class contains between 40 and 800 images. Optional annotations include bounding box coordinates and object contours.
+#'
+#' You can specify the type of target returned:
+#' - `"category"`: returns the class label (e.g., `"accordion"`).
+#' - `"annotation"`: returns a list with bounding box coordinates and object contour.
+#' - `"all"`: returns both the class label and the annotation.
+#'
+#' @param root Character. Root directory for dataset storage. Default is a temporary directory. Data will be stored under `root/caltech101`.
+#' @param target_type Character. One of `"category"`, `"annotation"`, or `"all"`. Determines the type of target returned. Default is `"category"`.
```
Comment (cregouby) [inline R/dataset-caltech.R:12]: **todo** please remove @param target_type. This is a detection dataset (This is my fault sorry. I just discover it. I've update the issue) and thus it requires both `labels` and `boxes`.

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** Delivery is fine for a classification dataset
**todo** I just discover from your delivery that it is a object detection dataset. So we have here to conform to the coco names, description, examples... I apologize for that.

File: R/dataset-caltech.R [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+#' first_item$y$box_coord      # Numeric vector: [x1, y1, x2, y2]
+#' first_item$y$obj_contour    # Matrix of contour points: N x 2
+#'
+#' # Category + Annotation target (All target)
+#' ds3 <- caltech101_dataset(target_type = "all", download = TRUE)
+#' first_item <- ds3[1]
+#' first_item$x  # RGB image array
+#' first_item$y$label          # e.g., "accordion"
+#' first_item$y$box_coord      # Numeric vector: [x1, y1, x2, y2]
+#' first_item$y$obj_contour    # Matrix of contour points: N x 2
```
Comment (cregouby) [inline R/dataset-caltech.R:43]: **todo** **missing** for object detection dataset, we want to see it, so we need a `bboxes <- draw_bounding_boxes(item$x, ...)` followed by a `tensor_image_browse(bboxes)` in the example.
**improvement** have a look at #170 and #171

File: R/dataset-caltech.R (hunk 2 above)
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:19]: Thank you for the improvement. I have updated the `x` to return array of shape `{C, H, W}`.

File: R/dataset-caltech.R [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    if (self$check_exists()) return()
+    rlang::inform("Downloading Caltech101 archive...")
+    fs::dir_create(self$root)
+    invisible(lapply(self$resources, function(res) {
+      zip_path <- download_and_cache(res$url, prefix = class(self)[1])
+      dest <- fs::path(self$root, fs::path_file(res$filename))
+      fs::file_copy(zip_path, dest, overwrite = TRUE)
+      md5_actual <- tools::md5sum(dest)[[1]]
+      if (md5_actual != res$md5) {
+        runtime_error(glue::glue("MD5 mismatch for file: {res$filename} (expected {res$md5}, got {md5_actual})"))
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:179]: I have used the following to maintain consistency, hope that's fine.
```r
runtime_error("Corrupt file! Delete the file in {archive} and try again.")
```

File: tests/testthat/test-dataset-caltech.R [hunk 6]
```diff
@@ -0,0 +1,117 @@
+context("dataset-caltech")
+
+test_that("Caltech101 dataset works correctly", {
+  t <- tempdir()
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply tests/testthat/test-dataset-caltech.R:4]: I have adopted a different method where it automatically gets deleted as you suggested previously.
```r
t <- withr::local_tempdir()
```

File: tests/testthat/test-dataset-caltech.R [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  expect_tensor_shape(batch$x,c(4,3,224,224))
+  expect_tensor_dtype(batch$x,torch_float())
+  expect_type(batch$y,"list")
+  expect_type(batch$y[[1]],"character")
+  expect_true(batch$y[[1]] %in% ds$classes)
+  expect_true(batch$y[[2]] %in% ds$classes)
+  expect_true(batch$y[[3]] %in% ds$classes)
+  expect_true(batch$y[[4]] %in% ds$classes)
+
+  unlink(file.path(t, "caltech-101"), recursive = TRUE)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply tests/testthat/test-dataset-caltech.R:68]: I have adopted a new method where it automatically gets deleted as you suggested previously.
```r
t <- withr::local_tempdir()
```

File: tests/testthat/test-dataset-caltech.R [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  expect_tensor_shape(batch$x,c(4,3,224,224))
+  expect_tensor_dtype(batch$x,torch_float())
+  expect_type(batch$y,"list")
+  expect_type(batch$y[[1]],"character")
+  expect_equal(batch$y[[1]],"ak47")
+  expect_equal(batch$y[[2]],"ak47")
+  expect_equal(batch$y[[3]],"ak47")
+  expect_equal(batch$y[[4]],"ak47")
+  
+  unlink(file.path(t, "caltech256"), recursive = TRUE)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply tests/testthat/test-dataset-caltech.R:116]: I have adopted a new method where it automatically gets deleted as you suggested previously.
```r
t <- withr::local_tempdir()
```

File: R/dataset-caltech.R (hunk 2 above)
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:19]: Thank you for the reference to coco documentation. 
```r
#' @return
#' A torch dataset. Each item is a list with two elements:
#' \describe{
#'   \item{x}{A 3D \code{torch_tensor} of shape \code{(3, H, W)} representing the image in RGB format.}
#'   \item{y}{A list with:
#'     \describe{
#'       \item{boxes}{A 2D \code{torch_tensor} of shape \code{(1, 4)} containing the bounding box in \code{xywh} format.}
#'       \item{labels}{A character scalar giving the class label for the image.}
#'       \item{contour}{A 2D \code{torch_tensor} of shape \code{(N, 2)} with the object contour coordinates.}
#'     }
#'   }
#' }
```
I have made it consistent with the coco documentation.

File: R/dataset-caltech.R (hunk 3 above)
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:12]: No worries at all. I've removed the @param `target_type` as requested and ensured the dataset now consistently returns both labels and boxes.Thanks for the clarification!

File: R/dataset-caltech.R (hunk 4 above)
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:43]: Thanks for the todo. I've added the `draw_bounding_boxes()` and `tensor_image_browse()` lines to the example to visualise the object detection output.

**Comment by DerrickUnleashed** [comment]
I’ve updated `_pkgdown.yml`, but I’m not entirely sure it aligns with what’s expected. Could you please take a look ?

File: tests/testthat/test-dataset-caltech.R [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  expect_true(batch$y[[1]] %in% ds$classes)
+  expect_true(batch$y[[2]] %in% ds$classes)
+  expect_true(batch$y[[3]] %in% ds$classes)
+  expect_true(batch$y[[4]] %in% ds$classes)
+
+  unlink(file.path(t, "caltech-101"), recursive = TRUE)
+})
+
+test_that("Caltech256 dataset works correctly", {
+  t <- tempdir()
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply tests/testthat/test-dataset-caltech.R:72]: I have adopted a new method where it automatically gets deleted as you suggested previously.
```r
t <- withr::local_tempdir()
```

**Comment by DerrickUnleashed** [comment]
I'll work on making sure that `Caltech256_dataset()` inherits from `Caltech101_dataset()`.

File: R/dataset-caltech.R
Comment (cregouby) [inline R/dataset-caltech.R:1]: **todo** don't use the `cli::cli_cli_inform()` but rather the local `cli_inform()` that provide translation

File: R/dataset-caltech.R [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    if (!fs::file_exists(ann_file))
+      runtime_error("Annotation file not found: {ann_file}")
+
+    if (!requireNamespace("R.matlab", quietly = TRUE))
+      runtime_error("Please install 'R.matlab' to read annotation files.")
+
+    mat_data <- R.matlab::readMat(as.character(ann_file))
+    coords <- as.numeric(mat_data[["box.coord"]])
+    box_coord <- torch_tensor(c(coords[1], coords[3], coords[2], coords[4]), dtype = torch_float())$unsqueeze(1)
+    box_coord <- box_convert(box_coord, in_fmt = "xywh", out_fmt = "xyxy")
```
Comment (cregouby) [inline R/dataset-caltech.R:124]: **praise** box_convert() is Nice and clear
**todo** your [["box_coord"]] manipulation seems over kill : 
```suggestion
boxes <- torch_tensor(mat_data[["box.coord"]])[,c(1,3,2,4)]
boxes <- box_convert(boxes, in_fmt = "xywh", out_fmt = "xyxy")
```

File: R/dataset-caltech.R [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+#'   image = first_item$x,
+#'   boxes = first_item$y$boxes,
+#'   labels = caltech101$classes[first_item$y$labels],
+#'   colors = "red"
+#' )
+#' tensor_image_browse(bboxes)
+#'
+#' # Draw contour points as keypoints
+#' contour_points <- draw_keypoints(
+#'   image = image,
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:40]: Oops! that should have slipped in by mistake. I'll use `first_item$x` as suggested. Thanks for pointing it out.

File: R/dataset-caltech.R [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+#'   radius = 2
+#' )
+#' tensor_image_browse(contour_points)
+#' }
+#'
+#' @name caltech101_dataset
+#' @aliases caltech101_dataset
+#' @title Caltech-101 Dataset
+#' @export
+caltech101_dataset <- dataset(
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:52]: Sure. I'll update it.

File: R/dataset-caltech.R [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    self$annotation_classes <- vapply(self$classes, function(x) if (x %in% names(name_map)) name_map[[x]] else x, character(1))
+
+    self$samples <- list()
+    self$labels <- c()
+    self$image_indices <- c()
+
+    for (i in seq_along(self$classes)) {
+      img_dir <- fs::path(obj_dir, self$classes[[i]])
+      imgs <- sort(fs::dir_ls(img_dir, glob = "*.jpg"))
+      self$samples <- append(self$samples, imgs)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:95]: Thanks. I'll keep in mind to not use `samples` as it may be misleading. I have updated it to `img_path`.

File: R/dataset-caltech.R
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:1]: Yes. I'll update to use internal `cli_` instead of `cli::cli_`.

File: R/dataset-caltech.R [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+#' first_item$x <- torch_tensor(first_item$x, dtype = torch::torch_uint8())
+#'
+#' # Draw bounding box
+#' bboxes <- draw_bounding_boxes(
+#'   image = first_item$x,
+#'   boxes = first_item$y$boxes,
+#'   labels = caltech101$classes[first_item$y$labels],
+#'   colors = "red"
+#' )
+#' tensor_image_browse(bboxes)
```
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:36]: I tried multiple methods the following made it much better
```r
mat_data <- R.matlab::readMat(as.character(ann_file))
boxes <- mat_data[["box.coord"]][c(1, 3, 2, 4)]
boxes <- matrix(boxes, nrow = 1)
boxes[, 3] <- boxes[, 3] * 2
boxes <- box_convert(torch_tensor(boxes), in_fmt = "xywh", out_fmt = "xyxy")
height <- dim(img)[2]
width <- dim(img)[3]
boxes <- clip_boxes_to_image(boxes, size = c(height, width))
```
What are your thoughts on it ?

[the remark this answers is not in the document]
Comment (cregouby) [reply R/dataset-caltech.R:36]: I like the straightforward `boxes <- torch_tensor(mat_data[["box.coord"]])[,c(1,3,2,4)]` but please don't try to do  `boxes[, 3] <- boxes[, 3] * 2`, nor `clip_boxes_to_image()` : that would completely messs the retraining of any model, and clip_box is intended for visualization, not for dataset modification...

**Comment by cregouby** [comment]
> 
> I find the `_pkgdown.yml` conflicts hard to comprehend.

I've made a draft publication[ in here](https://cregouby.github.io/torchvision/dev/reference/index.html) if it helps...

File: R/dataset-caltech.R (hunk 14 above)
[the remark this answers is not in the document]
Comment (DerrickUnleashed) [reply R/dataset-caltech.R:36]: Thanks for the suggestion. I'll go back to the straightforward method itself.

**Comment by DerrickUnleashed** [comment]
> > I find the `_pkgdown.yml` conflicts hard to comprehend.
> 
> I've made a draft publication[ in here](https://cregouby.github.io/torchvision/dev/reference/index.html) if it helps...

Thank you so much. I do now understand the complete structure of `_pkgdown.yml`.

**Review by cregouby — CHANGES_REQUESTED** [review]
**Praise** Thanks for all those improvements.
**todo** Having a look at the quality of detection values in caltech101 : 
image
It seem that we should'nt use it, and as other did (like[utltralytics](https://docs.ultralytics.com/fr/datasets/classify/caltech101/#how-can-i-train-an-ultralytics-yolo-model-on-the-caltech-101-dataset) , downgrade the dataset to a classification dataset. I'm sorry for this as it was my demand to switch to detection dataset. I sincerely apologize for this.

**Comment by DerrickUnleashed** [comment]
@cregouby I have made the changes necessary for `flowers102_dataset` in the `_pkgdown.yml` file...So, I'm hoping that commit gets pushed perfectly also.
